### PR TITLE
[SPARK-37090][BUILD] Upgrade `libthrift` to 0.16.0 to avoid security vulnerabilities

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR ported HIVE-21498, HIVE-25098 and upgraded libthrift to 0.16.0.

The CHANGES list for libthrift 0.16.0 is available at: https://github.com/apache/thrift/blob/v0.16.0/CHANGES.md

### Why are the changes needed?

To address [CVE-2020-13949](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-13949).

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @HyukjinKwon @juliuszsompolski We can upgrade to 0.13.0 first. If 0.16.0 is released, we can upgrade to 0.16.0 because we need [this patch](https://github.com/apache/thrift/pull/2470).

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
cc 
@srowen and @dongjoon-hyun too FYI

**Comment by dongjoon-hyun** [comment]
Let's wait for a while since this is not for Apache Spark 3.2.1.
To prevent accidental merging, I converted this PR to the draft, @wangyum .

**Review by srowen — COMMENTED** [review]
This seems OK if tests all pass this time

**Review by HyukjinKwon — APPROVED** [review]

**Review by srowen — COMMENTED** [review]
@wangyum if you're OK with this, you can mark it as ready to merge. Looks OK to me

**Comment by dongjoon-hyun** [comment]
I converted it back to `Ready for Review` because I did it before before Apache Spark 3.2.1 release.

**Review by dongjoon-hyun — COMMENTED** [review]
The last commit failed with the following. Could you re-trigger the tests, @wangyum ?
```
Error:  Failed to execute goal on project spark-hive_2.12: Could not resolve dependencies for project org.apache.spark:spark-hive_2.12:jar:spark-975225:
Could not find artifact org.apache.thrift:libthrift:jar:0.16.0 in gcs-maven-central-mirror (https://maven-central.storage-download.googleapis.com/maven2/) -> [Help 1]
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Pending CIs. And, please make the PR description up-to-date.
> This pr backport HIVE-21498 to upgrade libthrift to 0.13.0.

**Comment by srowen** [comment]
Ping @wangyum

**Comment by srowen** [comment]
I updated the description manually; I think this is good to go?

**Comment by dongjoon-hyun** [comment]
> I updated the description manually; I think this is good to go?

We have no evidence that this PS passes the CIs, don't we, @srowen ? As I mentioned here (https://github.com/apache/spark/pull/34362#pullrequestreview-867034822), the latest CI failed.

**Comment by srowen** [comment]
Ah I see the tests passing but did not all applicable tests run? Is it because the tests need to be enabled to run in the submitter's acct?

**Comment by dongjoon-hyun** [comment]
- No, no tests ran. It was a build error due to `Could not find artifact org.apache.thrift:libthrift:jar:0.16.0 ` as I wrote in the previous comment.
- When you cannot check the result in Apache Spark repo, you need to visit their branch because sometimes GitHub Action notification doesn't work properly. In this PR, the following is his branch. And, you can see the actual GitHub Action result at the commit.
    - https://github.com/wangyum/spark/tree/SPARK-37090

**Comment by wangyum** [comment]
Sorry. I was on Chinese new year holiday last 2 weeks. I used to test [Thrift 0.16-rc0](https://www.mail-archive.com/dev@thrift.apache.org/msg51912.html), Thrift 0.16 has not been officially released.

**Comment by dongjoon-hyun** [comment]
No problem. Please let us know when the PR is ready, @wangyum .

File: sql/hive/src/main/java/org/apache/thrift/transport/TFramedTransport.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+import org.apache.thrift.TByteArrayOutputStream;
+import org.apache.thrift.TConfiguration;
+
+/**
+ * This is based on libthrift-0.12.0 {@link org.apache.thrift.transport.TFramedTransport}.
+ * To fix class of org.apache.thrift.transport.TFramedTransport not found after upgrading libthrift.
+ *
+ * TFramedTransport is a buffered TTransport that ensures a fully read message
+ * every time by preceding messages with a 4-byte frame size.
+ */
```
Comment (wangyum) [inline sql/hive/src/main/java/org/apache/thrift/transport/TFramedTransport.java:30]: Origin Java class: https://github.com/apache/thrift/blob/0.12.0/lib/java/src/org/apache/thrift/transport/TFramedTransport.java

**Comment by srowen** [comment]
Looks like librthrift 0.16.0 was released and this passes, so should be good to go. Thoughts about backporting this to 3.2 and 3.1? (3.0, I presume, is EOL now). I guess I'm inclined to unless there is a non-trivial risk of breaking something, like Hadoop 2 compatibility

**Comment by wangyum** [comment]
This is for branch 3.2 and 3.1:
https://github.com/apache/spark/compare/branch-3.2...wangyum:SPARK-37090-branch-3.2?expand=1
https://github.com/apache/spark/compare/branch-3.1...wangyum:SPARK-37090-branch-3.1?expand=1

**Comment by srowen** [comment]
One last check from @dongjoon-hyun maybe; this and backports looks OK

**Comment by dongjoon-hyun** [comment]
Thank you for asking, @srowen . Sounds reasonable. No objection for backporting from my side.

**Comment by srowen** [comment]
Merged to master. I'll look at the backports next

**Comment by srowen** [comment]
Oh @wangyum do you have pull requests for the backports? the content looks OK, just bears testing

**Comment by dongjoon-hyun** [comment]
Hi, All. This is reverted from all branches due to the regression.
Please see https://github.com/apache/spark/pull/35646

**Pull request closed without merging**